api/filrewardsd/service/service.go

```diff
@@ -0,0 +1,571 @@
+	pb.RewardType_REWARD_TYPE_FIRST_BUCKET_ARCHIVE_CREATED: {factor: 2},
+	pb.RewardType_REWARD_TYPE_FIRST_MAILBOX_CREATED:        {factor: 1},
+	pb.RewardType_REWARD_TYPE_FIRST_THREAD_DB_CREATED:      {factor: 1},
+}
+
+type meta struct {
+	factor int32
+}
+
+type reward struct {
+	OrgKey            string        `bson:"org_key"`
+	DevKey            string        `bson:"dev_key"`
+	Type              pb.RewardType `bson:"type"`
+	Factor            int32         `bson:"factor"`
+	BaseAttoFILReward int32         `bson:"base_atto_fil_reward"`
+	CreatedAt         time.Time     `bson:"created_at"`
+}
+
+type claim struct {
+	OrgKey    string    `bson:"org_key"`
+	ClaimedBy string    `bson:"claimed_by"`
+	Amount    int32     `bson:"amount"`
+	CreatedAt time.Time `bson:"created_at"`
+}
+
+type orgKeyLock string
+
+func (l orgKeyLock) Key() string {
+	return string(l)
+}
+
+var _ nutil.SemaphoreKey = (*orgKeyLock)(nil)
+
+var _ pb.FilRewardsServiceServer = (*Service)(nil)
+
+type Service struct {
+	rewardsCol        *mongo.Collection
+	claimsCol         *mongo.Collection
+	ac                *analytics.Client
+	rewardsCacheOrg   map[string]map[pb.RewardType]struct{}
+	rewardsCacheDev   map[string]map[pb.RewardType]struct{}
+	baseAttoFILReward int32
+	server            *grpc.Server
+	semaphores        *nutil.SemaphorePool
+}
+
+type Config struct {
+	Listener          net.Listener
+	MongoUri          string
+	MongoDbName       string
+	AnalyticsAddr     string
+	BaseAttoFILReward int32
+	Debug             bool
+}
+
+func New(ctx context.Context, config Config) (*Service, error) {
+	if config.Debug {
+		if err := util.SetLogLevels(map[string]logging.LogLevel{
+			"filrewards": logging.LevelDebug,
+		}); err != nil {
+			return nil, err
+		}
+	}
+
+	client, err := mongo.Connect(ctx, options.Client().ApplyURI(config.MongoUri))
+	if err != nil {
+		return nil, fmt.Errorf("connecting to mongo: %v", err)
+	}
+	db := client.Database(config.MongoDbName)
+	rewardsCol := db.Collection(rewardsCollectionName)
+	claimsCol := db.Collection(claimsCollectionName)
+	_, err = rewardsCol.Indexes().CreateMany(ctx, []mongo.IndexModel{
+		{
+			Keys:    bson.D{primitive.E{Key: "org_key", Value: 1}, primitive.E{Key: "type", Value: 1}},
+			Options: options.Index().SetUnique(true).SetPartialFilterExpression(bson.M{"org_key": bson.M{"$gt": ""}}),
+		},
+		{
+			Keys:    bson.D{primitive.E{Key: "dev_key", Value: 1}, primitive.E{Key: "type", Value: 1}},
+			Options: options.Index().SetUnique(true).SetPartialFilterExpression(bson.M{"dev_key": bson.M{"$gt": ""}}),
+		},
+		{
+			Keys: bson.D{primitive.E{Key: "org_key", Value: 1}},
+		},
+		{
+			Keys: bson.D{primitive.E{Key: "dev_key", Value: 1}},
+		},
+		{
+			Keys: bson.D{primitive.E{Key: "type", Value: 1}},
+		},
+		{
+			Keys: bson.D{primitive.E{Key: "created_at", Value: 1}},
+		},
+	})
+	if err != nil {
+		return nil, fmt.Errorf("creating indexes: %v", err)
+	}
+
+	// Populate caches.
```

Comment:
@sanderpick and @textileben here are those in memory caches we were talking about. This is state that isn't in mongo. It is not required, but just a convenient/fast way to avoid reading from the database in certain cases.